### PR TITLE
New extractor for Qingting.fm(蜻蜓FM)

### LINKED ISSUE #5323: New extractor for Qingting.fm(蜻蜓FM)
### DO NOT REMOVE OR SKIP THE ISSUE TEMPLATE

- [X] I understand that I will be **blocked** if I remove or skip any mandatory\* field

### Checklist

- [X] I'm reporting a new site support request
- [X] I've verified that I'm running yt-dlp version **2022.10.04** ([update instructions](https://github.com/yt-dlp/yt-dlp#update)) or later (specify commit)
- [X] I've checked that all provided URLs are playable in a browser with the same IP and same login details
- [X] I've checked that none of provided URLs [violate any copyrights](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#is-the-website-primarily-used-for-piracy) or contain any [DRM](https://en.wikipedia.org/wiki/Digital_rights_management) to the best of my knowledge
- [X] I've searched the [bugtracker](https://github.com/yt-dlp/yt-dlp/issues?q=) for similar issues **including closed ones**. DO NOT post duplicates
- [X] I've read the [guidelines for opening an issue](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#opening-an-issue)
- [ ] I've read about [sharing account credentials](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#are-you-willing-to-share-account-details-if-needed) and am willing to share it if required

### Region

China

### Example URLs

- single audio: https://www.qingting.fm/channels/378005/programs/23023573
- single audio: https://m.qingting.fm/vchannels/378005/programs/23023573

### Provide a description that is worded well enough to be understood

Qingting.fm(蜻蜓FM) is a network audio application. Since its release in September 2011, it has grown into one of the leading audio content aggregation platforms in China. 

A new extractor for Qingting.fm(蜻蜓FM) will help listeners to download favorite audios and save them personally.

### Provide verbose output that clearly demonstrates the problem

- [X] Run **your** yt-dlp command with **-vU** flag added (`yt-dlp -vU <your command line>`)
- [X] Copy the WHOLE output (starting with `[debug] Command-line config`) and insert it below

### Complete Verbose Output

```shell
[debug] Command-line config: ['-vU', 'https://www.qingting.fm/channels/378005/programs/23023573']                              
[debug] Encodings: locale UTF-8, fs utf-8, pref UTF-8, out utf-8, error utf-8, screen utf-8                                    
[debug] yt-dlp version 2022.10.04 [4e0511f] (pip) API                                                                                    
[debug] Python 3.10.8 (CPython 64bit) - macOS-11.6-x86_64-i386-64bit                                                           
[debug] Checking exe version: ffmpeg -bsfs                                                                                     
[debug] Checking exe version: ffprobe -bsfs                                                                                    
[debug] exe versions: ffmpeg 5.1.2 (setts), ffprobe 5.1.2, rtmpdump 2.4                                                        
[debug] Optional libraries: Cryptodome-3.15.0, brotli-1.0.9, certifi-2022.09.24, mutagen-1.45.1, sqlite3-2.6.0, websockets-10.3
[debug] Proxy map: {}                                                                                                          
[debug] Loaded 1690 extractors                                                                                                 
[debug] Fetching release info: https://api.github.com/repos/yt-dlp/yt-dlp/releases/latest                                      
Latest version: 2022.10.04, Current version: 2022.10.04                                                                        
yt-dlp is up to date (2022.10.04)                                                                                              
[debug] [generic] Extracting URL: https://www.qingting.fm/channels/378005/programs/23023573                                    
[generic] 23023573: Downloading webpage                                                                                        
[redirect] Following redirect to https://www.qingting.fm/channels/378005/programs/23023573/                                    
[debug] [generic] Extracting URL: https://www.qingting.fm/channels/378005/programs/23023573/                                   
[generic] 23023573: Downloading webpage                                                                                        
WARNING: [generic] Falling back on generic information extractor                                                               
[generic] 23023573: Extracting information                                                                                     
[debug] Looking for Brightcove embeds                                                                                          
[debug] Looking for embeds                                                                                                     
ERROR: Unsupported URL: https://www.qingting.fm/channels/378005/programs/23023573/                                             
Traceback (most recent call last):                                                                                             
  File "/usr/local/Cellar/yt-dlp/2022.10.4/libexec/lib/python3.10/site-packages/yt_dlp/YoutubeDL.py", line 1477, in wrapper       
    return func(self, *args, **kwargs)                                                                                                   
  File "/usr/local/Cellar/yt-dlp/2022.10.4/libexec/lib/python3.10/site-packages/yt_dlp/YoutubeDL.py", line 1553, in __extract_info
    ie_result = ie.extract(url)                                                                                                          
  File "/usr/local/Cellar/yt-dlp/2022.10.4/libexec/lib/python3.10/site-packages/yt_dlp/extractor/common.py", line 672, in extract       
    ie_result = self._real_extract(url)                                                                                                  
  File "/usr/local/Cellar/yt-dlp/2022.10.4/libexec/lib/python3.10/site-packages/yt_dlp/extractor/generic.py", line 3062, in _real_extract 
    raise UnsupportedError(url)                                                                                                
yt_dlp.utils.UnsupportedError: Unsupported URL: https://www.qingting.fm/channels/378005/programs/23023573/
```

### PR DESCRIPTION
**Pull request by changren-wcr**
**IMPORTANT**: PRs without the template will be CLOSED

### Description of your *pull request* and other information

</details>

ADD DESCRIPTION HERE

Fixes #5323

<details open><summary>Template</summary> 

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [contributing guidelines](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions) including [yt-dlp coding conventions](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#yt-dlp-coding-conventions)
- [x] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8) and [ran relevant tests](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions)

### In order to be accepted and merged into yt-dlp each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [ ] Fix or improvement to an extractor (Make sure to add/update tests)
- [x] New extractor ([Piracy websites will not be accepted](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#is-the-website-primarily-used-for-piracy))
- [ ] Core bug fix/improvement
- [ ] New feature (It is strongly [recommended to open an issue first](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#adding-new-feature-or-making-overarching-changes))

### CODE REVIEW

**Review by bashonly — CHANGES_REQUESTED** [review]
this needs some work. I've suggested changes to:
 - make the extractor adhere to yt-dlp coding conventions
 - bypass downloading the desktop webpage altogether
 - properly parse the mobile webpage's JSON and extract info

File: yt_dlp/extractor/qingting.py [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+    _VALID_URL = r'https?://(?:www\.|m\.)?(?:qingting\.fm|qtfm\.cn)/v?channels/(?P<channel>\d+)/programs/(?P<id>\d+)'
+    _TESTS = {
+        'url': 'https://www.qingting.fm/channels/378005/programs/22257411/',
+        'md5': '47e6a94f4e621ed832c316fd1888fb3c',
+        'info_dict': {
+            'id': '22257411',
+            'ext': 'mp3',
+            'title': '用了十年才修改，谁在乎教科书？-睡前消息-蜻蜓FM听头条',
+        }
+    }
```
Comment (pukkandan) [inline yt_dlp/extractor/qingting.py:16]:
```suggestion
    _TESTS = [{
        'url': 'https://www.qingting.fm/channels/378005/programs/22257411/',
        'md5': '47e6a94f4e621ed832c316fd1888fb3c',
        'info_dict': {
            'id': '22257411',
            'ext': 'mp3',
            'title': '用了十年才修改，谁在乎教科书？-睡前消息-蜻蜓FM听头条',
        }
    }]
```

File: yt_dlp/extractor/qingting.py [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+        channel_id, pid = self._match_valid_url(url).groups()
+        webpage = self._download_webpage(
+            f'https://m.qtfm.cn/vchannels/{channel_id}/programs/{pid}/', pid)
+        info = self._search_json(r'window\.__initStores\s*=', webpage, 'program info', pid)
+        return {
+            'id': pid,
+            'title': traverse_obj(info, ('ProgramStore', 'programInfo', 'title')),
+            'channel_id': channel_id,
+            'url': traverse_obj(info, ('ProgramStore', 'programInfo', 'audioUrl')),
+            'ext': 'm4a',
```
Comment (pukkandan) [inline yt_dlp/extractor/qingting.py:28]:
```suggestion
            'vcodec': 'none',
            'acode': 'm4a',
            'ext': 'm4a',
```

**Comment by changren-wcr** [comment]
@bashonly @pukkandan Thanks for your patience and kindness! All the problems have been solved.

File: yt_dlp/extractor/qingting.py [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+            'title': '【睡前消息488】重庆山火之后，有图≠真相',
+            'channel_id': '378005',
+            'channel_name': '睡前消息',
+            'podcaster': '马督工',
+            'ext': 'm4a',
+        }
+    }]
+
+    def _real_extract(self, url):
+        channel_id, pid = self._match_valid_url(url).groups()
```
Comment (pukkandan) [inline yt_dlp/extractor/qingting.py:33]:
```suggestion
        channel_id, pid = self._match_valid_url(url).group('channel', 'id')
```

File: yt_dlp/extractor/qingting.py [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+        web_page = self._download_webpage(
+            f'https://m.qtfm.cn/vchannels/{channel_id}/programs/{pid}/', pid)
+        info = self._search_json(r'window\.__initStores\s*=', web_page, 'program info', pid)
```
Comment (pukkandan) [inline yt_dlp/extractor/qingting.py:36]:
```suggestion
        webpage = self._download_webpage(
            f'https://m.qtfm.cn/vchannels/{channel_id}/programs/{pid}/', pid)
        info = self._search_json(r'window\.__initStores\s*=', webpage, 'program info', pid)
```

**Review by pukkandan — APPROVED** [review]

File: yt_dlp/extractor/qingting.py [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+        channel_id, pid = self._match_valid_url(url).group('channel', 'id')
+        webpage = self._download_webpage(
+            f'https://m.qtfm.cn/vchannels/{channel_id}/programs/{pid}/', pid)
+        info = self._search_json(r'window\.__initStores\s*=', webpage, 'program info', pid)
+        return {
+            'id': pid,
+            'title': traverse_obj(info, ('ProgramStore', 'programInfo', 'title')),
+            'channel_id': channel_id,
+            'channel_name': traverse_obj(info, ('ProgramStore', 'channelInfo', 'title')),
+            'podcaster': traverse_obj(info, ('ProgramStore', 'podcasterInfo', 'podcaster', 'nickname')),
```
Comment (bashonly) [inline yt_dlp/extractor/qingting.py:42]:
```suggestion
            'channel': traverse_obj(info, ('ProgramStore', 'channelInfo', 'title')),
            'uploader': traverse_obj(info, ('ProgramStore', 'podcasterInfo', 'podcaster', 'nickname')),
```

File: yt_dlp/extractor/qingting.py [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+    _VALID_URL = r'https?://(?:www\.|m\.)?(?:qingting\.fm|qtfm\.cn)/v?channels/(?P<channel>\d+)/programs/(?P<id>\d+)'
+    _TESTS = [{
+        'url': 'https://www.qingting.fm/channels/378005/programs/22257411/',
+        'md5': '47e6a94f4e621ed832c316fd1888fb3c',
+        'info_dict': {
+            'id': '22257411',
+            'title': '用了十年才修改，谁在乎教科书？',
+            'channel_id': '378005',
+            'channel_name': '睡前消息',
+            'podcaster': '马督工',
```
Comment (bashonly) [inline yt_dlp/extractor/qingting.py:16]:
```suggestion
            'channel': '睡前消息',
            'uploader': '马督工',
```

File: yt_dlp/extractor/qingting.py [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+        }
+    }, {
+        'url': 'https://m.qtfm.cn/vchannels/378005/programs/23023573/',
+        'md5': '2703120b6abe63b5fa90b975a58f4c0e',
+        'info_dict': {
+            'id': '23023573',
+            'title': '【睡前消息488】重庆山火之后，有图≠真相',
+            'channel_id': '378005',
+            'channel_name': '睡前消息',
+            'podcaster': '马督工',
```
Comment (bashonly) [inline yt_dlp/extractor/qingting.py:27]:
```suggestion
            'channel': '睡前消息',
            'uploader': '马督工',
```

**Review by bashonly — APPROVED** [review]

**Review by bashonly — APPROVED** [review]

**Pull request merged**